enterprise/server/backends/migration_cache/migration_cache.go

```diff
@@ -0,0 +1,157 @@
+package migration_cache
+
+import (
+	"context"
+	"flag"
+	repb "github.com/buildbuddy-io/buildbuddy/proto/remote_execution"
+	"github.com/buildbuddy-io/buildbuddy/server/environment"
+	"github.com/buildbuddy-io/buildbuddy/server/interfaces"
+	"github.com/buildbuddy-io/buildbuddy/server/util/alert"
+	"github.com/buildbuddy-io/buildbuddy/server/util/log"
+	"golang.org/x/sync/errgroup"
+	"io/fs"
+	"math/rand"
+	"os"
+	"path/filepath"
+	"sync"
+)
+
+var (
+	doubleReadPercentage      = flag.Float64("cache.migration_cache.double_read_percentage", 0, "The percentage of reads we should double read, to ensure the two caches have the same data. Float [0, 1]. 0 for no double reads, 1 for double reads on every read")
+	migrateAndExit            = flag.Bool("cache.migration_cache.migrate_and_exit", false, "If true, attempt to migrate src cache to dest cache and exit upon completion.")
+	clearCacheBeforeMigration = flag.Bool("cache.migration.clear_cache_before_migration", false, "If set, clear any existing cache content in the dest cache.")
+	autoFixDoubleReadErrs     = flag.Bool("cache.migration.auto_fix_double_read_errs", false, "If set, automatically fix any double read errs in the dest cache")
+)
+
+type MigrationCache struct {
+	Src        interfaces.Cache
+	Dest       MigrationDestCache
+	SrcRootDir string
+
+	mu             sync.RWMutex
+	eg             *errgroup.Group
+	doubleReadChan chan *repb.Digest
+}
+
+type MigrationDestCache interface {
+	interfaces.Cache
+
+	// Implementation should be async, so that the double write doesn't slow down operations on the primary db
```

Comment:
With double reads and double writes -- one way I've seen it done is that both take place in parallel, synchronously in the request, and then if the result differs for some reason, a warning is logged. 

If they happen async this isn't possible. What are the tradeoffs involved with doing this async vs sync?

Reply:
Don't know if you saw ProcessDoubleReads() but thought that could be a good way to have sync double reads running in the background, without affecting performance of Get() on the src db

As for double writes: 
Async
+ Performance of source db is not affected. Ideally customers should not know migration is happening, if there's a performance problem with writes to the dest db
- Harder to debug if there's some lag propagating changes to dest db
- Depending on implementation of async write log, might lose data when server restarts. Could be messy to fix those discrepancies
- Switching over source of truth to destination db might be a little off if not all the updates have completed double writing (though probably minor)

Sync
+ Simplicity
- If the writes run in parallel and the src write fails but dest write succeeds, will need to "rollback" changes to the dest bc the src db should be the source of truth. Will need an additional sync delete on the dest db
- If there's a performance problem writing to the new db, could cause propagating performance problems bc we need to hold the lock when writing to both dbs

Listing them out, sync approach might actually be better. We can add another flag to only double write a proportion of writes and slowly ramp up, so we can identify and fix any double write performance problems before affecting all write traffic

Reply:
Don't know if you saw ProcessDoubleReads() but thought that could be a good way to have sync double reads running in the background, without affecting performance of Get() on the src db

As for double writes:
Async
/+ Performance of source db is not affected. Ideally customers should not know migration is happening, if there's a performance problem with writes to the dest db
/- Harder to debug if there's some lag propagating changes to dest db
/- Depending on implementation of async write log, might lose data when server restarts. Could be messy to fix those discrepancies
/- Switching over source of truth to destination db might be a little off if not all the updates have completed double writing (though probably minor)

Sync
/+ Simplicity
/- If the writes run in parallel and the src write fails but dest write succeeds, will need to "rollback" changes to the dest bc the src db should be the source of truth. Will need an additional sync delete on the dest db
/- If there's a performance problem writing to the new db, could cause propagating performance problems bc we need to hold the lock when writing to both dbs

Listing them out, sync approach might actually be better. We can add another flag to only double write a proportion of writes and slowly ramp up, so we can identify and fix any double write performance problems before affecting all write traffic

Reply:
I like the simplicity of sync.

One idea @vadimberezniker suggested was we could also enqueue stuff for migration when it is read

Reply:
Updated the proposal to copy data asynchronously during reads, as opposed to trying to copy everything from the src cache on startup. However that means we will need to do async double writes, otherwise we can't guarantee that the ordering will be correct between the copying and new writes